src/configPolling.ts

```diff
@@ -0,0 +1,144 @@
+import fetch from 'node-fetch';
+
+import ConfigService from './config';
+import {logger} from './logging';
+import {PollingConfig} from './types';
+import {validateConfig} from './validate';
+
+type StartOptions = {
+  /**
+   * Immediately trigger a polling tick when starting
+   */
+  immediate: boolean;
+};
+
+/**
+ * The ConfigPoller is used for two things:
+ *
+ *  1. During boot we poll on the bootInterval to load the configuration module.
+ *     Typically this will be a fairly quick poll to boot the service ASAP.
+ *
+ *  2. After it's booted, it will switch to the idleInterval. This is typically
+ *     much slower and is intended for the service to keep in sync with updates
+ *     to the configuration.
+ *
+ * It will automatically switch from the bootInterval to idleInterval after the
+ * first configuration module is loaded.
+ */
+export default class ConfigPoller {
+  /**
+   * The ConfigService the ConfigPoller is associated to
+   */
+  #config: ConfigService;
+  /**
+   * Configuration for polling
+   */
+  #settings: PollingConfig;
+  /**
+   * Tracks reference handle of the setInterval
+   */
+  #interval: NodeJS.Timeout | null = null;
+
+  constructor(config: ConfigService, settings: PollingConfig) {
+    this.#config = config;
+    this.#settings = settings;
+  }
+
+  /**
+   * The polling interval is dependant on if we've already successfully loaded a
+   * configuration module or not.
+   */
+  get pollInterval() {
+    return this.#config.isLoaded
+      ? this.#settings.idleInterval
+      : this.#settings.bootInterval;
+  }
+
+  /**
+   * Fired when the polling interval ticks
+   */
+  #pollTick = async () => {
+    logger.debug('Polling tick...');
+
+    // Load deadline should match the polling interval
+    const loadDeadline = this.pollInterval;
```

Comment:
See my `promiseWithTimeout` abstraction above. It makes even more sense to me now 🙂